xbmc/guilib/GUIVisualisationControl.cpp

```diff
@@ -19,76 +19,129 @@
+{
+  return m_pBuffer;
+}
+
+void CAudioBuffer::Set(const float* psBuffer, int iSize)
+{
+  if (iSize < 0)
+    return;
+
+  memcpy(m_pBuffer, psBuffer, iSize * sizeof(float));
+  for (int i = iSize; i < m_iLen; ++i)
+    m_pBuffer[i] = 0;
+}
+
 CGUIVisualisationControl::CGUIVisualisationControl(int parentID, int controlID, float posX, float posY, float width, float height)
-    : CGUIRenderingControl(parentID, controlID, posX, posY, width, height), m_bAttemptedLoad(false)
+  : CGUIControl(parentID, controlID, posX, posY, width, height),
+    m_callStart(false),
+    m_alreadyStarted(false),
+    m_attemptedLoad(false),
+    m_updateTrack(false),
+    m_instance(nullptr)
 {
   ControlType = GUICONTROL_VISUALISATION;
 }
 
 CGUIVisualisationControl::CGUIVisualisationControl(const CGUIVisualisationControl &from)
-  : CGUIRenderingControl(from), m_bAttemptedLoad(false), m_addon()
+  : CGUIControl(from),
+    m_callStart(false),
+    m_alreadyStarted(false),
+    m_attemptedLoad(false),
+    m_instance(nullptr)
 {
   ControlType = GUICONTROL_VISUALISATION;
 }
 
+std::string CGUIVisualisationControl::Name()
+{
+  if (m_instance == nullptr)
+    return "";
+  return m_instance->Name();
+}
+
 bool CGUIVisualisationControl::OnMessage(CGUIMessage &message)
 {
-  switch (message.GetMessage())
+  if (m_alreadyStarted)
   {
-  case GUI_MSG_GET_VISUALISATION:
-    message.SetPointer(m_addon.get());
-    return m_addon != NULL;
-  case GUI_MSG_VISUALISATION_RELOAD:
-    FreeResources(true);
-    return true;
-  case GUI_MSG_PLAYBACK_STARTED:
-    if (m_addon)
+    switch (message.GetMessage())
     {
-      m_addon->UpdateTrack();
+    case GUI_MSG_GET_VISUALISATION:
+      message.SetPointer(this);
+      return true;
+    case GUI_MSG_VISUALISATION_RELOAD:
+      FreeResources(true);
+      return true;
+    case GUI_MSG_PLAYBACK_STARTED:
+      m_updateTrack = true;
       return true;
+    default:
+      break;
     }
-    break;
   }
-  return CGUIRenderingControl::OnMessage(message);
+  return CGUIControl::OnMessage(message);
 }
 
 bool CGUIVisualisationControl::OnAction(const CAction &action)
 {
-  if (!m_addon)
-    return false;
-
-  switch (action.GetID())
+  if (m_alreadyStarted)
   {
-  case ACTION_VIS_PRESET_NEXT:
-    return m_addon->OnAction(VIS_ACTION_NEXT_PRESET);
-  case ACTION_VIS_PRESET_PREV:
-    return m_addon->OnAction(VIS_ACTION_PREV_PRESET);
-  case ACTION_VIS_PRESET_RANDOM:
-    return m_addon->OnAction(VIS_ACTION_RANDOM_PRESET);
-  case ACTION_VIS_RATE_PRESET_PLUS:
-    return m_addon->OnAction(VIS_ACTION_RATE_PRESET_PLUS);
-  case ACTION_VIS_RATE_PRESET_MINUS:
-    return m_addon->OnAction(VIS_ACTION_RATE_PRESET_MINUS);
-  case ACTION_VIS_PRESET_LOCK:
-    return m_addon->OnAction(VIS_ACTION_LOCK_PRESET);
-  default:
-    return CGUIRenderingControl::OnAction(action);
+    switch (action.GetID())
+    {
+    case ACTION_VIS_PRESET_NEXT:
+      m_instance->OnAction(VIS_ACTION_NEXT_PRESET, nullptr);
```

Comment:
This is exactly what I ment some time ago when I was saying that the speed of kodi-agile development will not pay out, because at the end there is double effort to port changes to master. :-/